File: src/test/java/sirius/kernel/commons/BOMReaderTest.java
/*
 * Made with all the love in the world
 * by scireum in Remshalden, Germany
 *
 * Copyright by scireum GmbH
 * http://www.scireum.de - info@scireum.de
 */

package sirius.kernel.commons;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.nio.charset.StandardCharsets;

class BOMReaderTest {

    private static final byte[] WITH_UTF8_BOM = {(byte) 239, (byte) 187, (byte) 191, 'H', 'E', 'L', 'L', 'O'};

    @Test
    void readBOM() throws IOException {
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        Writer writer = new OutputStreamWriter(outputStream, StandardCharsets.UTF_8);
        writer.write(Streams.UNICODE_BOM_CHARACTER);
        writer.write("HELLO");
        writer.flush();
        BOMReader in = new BOMReader(new InputStreamReader(new ByteArrayInputStream(outputStream.toByteArray())));
        Assertions.assertEquals('H', in.read());
        Assertions.assertEquals('E', in.read());
    }

    @Test
    void readWithoutBOM() throws IOException {
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        Writer writer = new OutputStreamWriter(outputStream, StandardCharsets.UTF_8);
        writer.write("HELLO");
        writer.flush();
        BOMReader in = new BOMReader(new InputStreamReader(new ByteArrayInputStream(outputStream.toByteArray())));
        Assertions.assertEquals('H', in.read());
        Assertions.assertEquals('E', in.read());
    }

    @ParameterizedTest
    @ValueSource(ints = {1, 2, 5})
    void readArrayBOM(int length) throws IOException {
        BOMReader in = new BOMReader(new InputStreamReader(new ByteArrayInputStream(WITH_UTF8_BOM)));
        char[] buf = new char[length];
        Assertions.assertEquals(length, in.read(buf));
        Assertions.assertEquals('H', buf[0]);
    }

    @Test
    void readArrayWithoutBOM() throws IOException {
        BOMReader in = new BOMReader(new InputStreamReader(new ByteArrayInputStream(WITH_UTF8_BOM)));
        char[] buf = new char[2];
        Assertions.assertEquals(2, in.read(buf));
        Assertions.assertEquals('H', buf[0]);
    }
}

File: src/test/java/sirius/kernel/commons/BOMReaderTest.kt
/*
 * Made with all the love in the world
 * by scireum in Remshalden, Germany
 *
 * Copyright by scireum GmbH
 * http://www.scireum.de - info@scireum.de
 */
package sirius.kernel.commons

import org.junit.jupiter.api.Test
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStreamReader
import kotlin.test.assertEquals

class BOMReaderTest {

    companion object {
        private val WITH_UTF8_BOM = byteArrayOf(
            239.toByte(),
            187.toByte(),
            191.toByte(),
            'H'.toByte(),
            'E'.toByte(),
            'L'.toByte(),
            'L'.toByte(),
            'O'.toByte()
        )
        private val WITHOUT_BOM = byteArrayOf('H'.toByte(), 'E'.toByte(), 'L'.toByte(), 'L'.toByte(), 'O'.toByte())
    }

    @Test
    @Throws(IOException::class)
    fun readBOM() {
        val reader = BOMReader(InputStreamReader(ByteArrayInputStream(WITH_UTF8_BOM)))
        assertEquals('H'.code.toLong(), reader.read().toLong())
        assertEquals('E'.code.toLong(), reader.read().toLong())
    }

    @Test
    @Throws(IOException::class)
    fun readWithoutBOM() {
        val reader = BOMReader(InputStreamReader(ByteArrayInputStream(WITHOUT_BOM)))
        assertEquals('H'.code.toLong(), reader.read().toLong())
        assertEquals('E'.code.toLong(), reader.read().toLong())
    }

    @Test
    @Throws(IOException::class)
    fun readArray1BOM() {
        val reader = BOMReader(InputStreamReader(ByteArrayInputStream(WITH_UTF8_BOM)))
        val buffer = CharArray(1)
        assertEquals(1, reader.read(buffer).toLong())
        assertEquals('H'.code.toLong(), buffer[0].code.toLong())
    }

    @Test
    @Throws(IOException::class)
    fun readArray2BOM() {
        val reader = BOMReader(InputStreamReader(ByteArrayInputStream(WITH_UTF8_BOM)))
        val buffer = CharArray(2)
        assertEquals(2, reader.read(buffer).toLong())
        assertEquals('H'.code.toLong(), buffer[0].code.toLong())
    }

    @Test
    @Throws(IOException::class)
    fun readArray10BOM() {
        val reader = BOMReader(InputStreamReader(ByteArrayInputStream(WITH_UTF8_BOM)))
        val buffer = CharArray(10)
        assertEquals(5, reader.read(buffer).toLong())
        assertEquals('H'.code.toLong(), buffer[0].code.toLong())
    }

    @Test
    @Throws(IOException::class)
    fun readArrayWithoutBOM() {
        val reader = BOMReader(InputStreamReader(ByteArrayInputStream(WITH_UTF8_BOM)))
        val buffer = CharArray(2)
        assertEquals(2, reader.read(buffer).toLong())
        assertEquals('H'.code.toLong(), buffer[0].code.toLong())
    }
}
